Issue 6 use npm

skymandr opened the pull request:
This PR is currently only a tool for me to get an overview. No need to review it, though of course, if you see anything dreadful do let me know.

Will eventually hopefully resolve #6 (will also be a majors step towards #62 and may resolve it, but that isn't guaranteed).

local-minimum commented on scanomatic/ui_server_data/compile.html:
```diff
@@ -51,14 +47,14 @@ <h1>Compile Project</h1>
+              onclick="som.setOnAllImages(true);"/>
             <input type="button" id="manual-select-none" value="Select None"
-              onclick="setOnAllImages(false);"/>
+              onclick="som.etOnAllImages(false);"/>
```
typo?
```suggestion
              onclick="som.setOnAllImages(false);"/>
```

skymandr replied:
indeed.

local-minimum commented on scanomatic/ui_server_data/experiment.html:
```diff
@@ -29,9 +23,9 @@ <h2>Metadata</h2>
-                onchange="set_experiment_root(this); validate_experiment();"
-                onfocus="set_experiment_root(this);"
-                onkeyup="set_experiment_root(this);"/>
+                onchange="som.setExperimentRoot(this); validateExperiment();"
```
```suggestion
                onchange="som.setExperimentRoot(this); som.validateExperiment();"
```
missing?

skymandr replied:
yep.

local-minimum commented on scanomatic/ui_server_data/experiment.html:
```diff
@@ -49,7 +43,7 @@ <h2>Metadata</h2>
             <td colspan="5">
               <textarea id="project-description" name="project-description"
-                onchange="validate_experiment();"></textarea>
+                onchange="validateExperiment();"></textarea>
```
```suggestion
                onchange="som.validateExperiment();"></textarea>
```

local-minimum commented on scanomatic/ui_server_data/fixture.html:
```diff
@@ -84,28 +79,28 @@ <h2 id="selected-fixture-name"></h2>
       <div>
         <button type="button" id="clear-areas"
-            onclick="clearAreas();draw_fixture();selected_fixture_canvas_jq.focus();">
+            onclick="som.clearAreas();som.drawFixture();selectedFixtureCanvasJq.focus();">
```
```suggestion
            onclick="som.clearAreas();som.drawFixture();som.selectedFixtureCanvasJq.focus();">
```

local-minimum replied:
?

skymandr replied:
Hmm... this is a global variable defined in `fixture.js`:
```javascript
let selectedFixtureCanvasJq;
```
I'll have to figure out how to handle that... might be a few (or a bunch) more that I have missed...

skymandr replied:
I think this is fixed now.

local-minimum replied:
looks reasonable here now

local-minimum commented on scanomatic/ui_server_data/js/som/analysis.js:
```diff
@@ -20,62 +46,52 @@ function set_fixture_plate_listing() {
     }
   } else {
-    fixt = $(current_fixture_id).val();
+    const fixt = $(currentFixtureId).val();
```
I think this might be a bit of a confusion. I saw that this was defined in the global scope in the html-code in various places but now it's defined here in the module... it should probably be either.

local-minimum replied:
There were some cases of this in various html-files

skymandr replied:
yes, probably same case as `selectedFixtureCanvasJq` above... Well spotted.

skymandr replied:
I think I've fixed all of them.

local-minimum reviewed (changes requested):
I'm not focused enough to fully review all the js-changes today.

local-minimum commented on package.json:
```diff
@@ -5,6 +5,8 @@
     "lint": "eslint --ext mjs,jsx,js scanomatic/ui_server_data/js/ccc scanomatic/ui_server_data/js/specs"
   },
   "devDependencies": {
+    "@types/spin": "^2.3.30",
```
I'm not sure, but i thought all `@types` were basically for typescript

skymandr replied:
It was just VSC complaining:
```
Could not find a declaration file for module 'spin'. '/home/skymandr/Work/SoM/scanomatic-standalone/node_modules/spin/spin.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/spin` if it exists or add a new declaration (.d.ts) file containing `declare module 'spin';`ts(7016)
  ```
  Not sure why it bothers telling me this when it is not a `.ts` file but oh well. Removed.

skymandr commented on scanomatic/ui_server_data/js/som/fixtures.js:
```diff
@@ -373,17 +351,16 @@ function drawHoverSlice(imageCoords) {
-  selectedFixtureCanvasJq = $(selectedFixtureCanvasId);
-  selectedFixtureCanvasJq.attr('tabindex', '0');
-  [selectedFixtureCanvas] = selectedFixtureCanvasJq;
+  const canvas = $(getSharedValue('selectedFixtureCanvasId'));
```
This function required the most rewriting to get it to work with setters/getters. I think I got it right, but please check.

skymandr commented on scanomatic/ui_server_data/compile.html:
```diff
@@ -86,11 +82,11 @@ <h3>Chaining</h3>
   <script>
     image_list_div = $('#manual-selection-div');
     manual_btn = $('#manual-selection');
```
Yes, I suppose so. I can add them.

skymandr replied:
`manual_btn` is only accessed locally, but I have migrated `image_list_div`.

skymandr replied:
Fixed in next push.

skymandr commented on scanomatic/ui_server_data/js/som/qc_normAPIHelper.js:
```diff
@@ -1,50 +1,57 @@
-        projects.push(project);
+    if (error) {
+      callback(null);
+      return console.warn(error);
```
Sure. Can I then add a lint-directive to allow console warn?

local-minimum replied:
yes

skymandr replied:
fixed in next push

skymandr commented on scanomatic/ui_server_data/js/som/helpers.js:
```diff
@@ -178,3 +192,8 @@ function setVersionInformation(target, preface) {
+const STORE = {};
+
+export const setSharedValue = (key, value) => { STORE[key] = value; };
+export const getSharedValue = key => STORE[key];
```
By warn, do you mean `console.warn`?

local-minimum replied:
yes

skymandr replied:
fixed in next push

skymandr commented on scanomatic/ui_server_data/js/som/analysis.js:
```diff
@@ -124,74 +131,79 @@ function get_regridding_setting(i) {
-function get_dir() {
-  return $('#compilation').val().replace(/\/[^\/]*$/, '');
+function regriddingSettingsData() {
+  const max = Math.max(...gridplates);
```
Fixed in next push.

skymandr commented on scanomatic/ui_server_data/js/som/qc_normAPIHelper.js:
```diff
@@ -1,50 +1,57 @@
 const branchSymbol = '¤';
 
-let lock;
+// let lock;
```
Fixed in next push.

local-minimum commented:
Errors on Compile view:
- CSS
![image](https://user-images.githubusercontent.com/8041100/149498921-3d5af5c5-108a-4657-b322-ddfb7d57b18e.png)

- Console has a lot of errors
- Noting happens on submit

local-minimum commented:
Errors on on Scan view:

- focus to Description field cause console error
- Age change causes console errors
- Time from pinning to experiment causes console errors
- Doesn't show CCC calibration, analysis page does so should not be an issue

local-minimum commented:
Analysis Page:
- No message on submit analysis (not chained)

local-minimum commented:
Feature Extraction
- Erasing `/root` text in analysis folder causes console error

local-minimum commented:
QC (Can't check much since):
- Uncaught TypeError: som.Spinner is not a constructor

local-minimum commented:
Fixtures:
- General CSS issues
- Uncaught TypeError: som.setsharedvalue is not a functio
- Uncaught TypeError: _jquery2.default(...)[0] is undefined attempting to detect markers on fixture

skymandr commented on scanomatic/ui_server_data/js/som/qc_normAPIHelper.js:
```diff
@@ -246,43 +221,126 @@ function GetYieldPlateData(url, placeholder, key, isNormalized, callback) {
-    callback(json);
+    if (json.success === false) {
+      alert(`Could not display the data! \n${json.reason}`);
+      callback(null);
```
fixed in next push

local-minimum commented on scanomatic/ui_server_data/js/som/analysis.js:
```diff
@@ -85,33 +101,24 @@ function loadgridimage(i) {
     $('#manual-regridding-image').append("<p class='error-message'>Could not find the grid image! Maybe gridding failed last time?</p>");
   }).always(() => {
-    if (show_gridimage) {
+    if (showGridImage) {
       $('#manual-regridding-image').show();
     } else {
       $('#manual-regridding-image').hide();
     }
```
always in asynchroneus so if hideGridImage is called before it happens it may not be true

skymandr replied:
So, what should be done? Is calling `hideGridImage` in the `else` acceptable?

skymandr replied:
I did that.

local-minimum replied:
As it is here now i good I hope

skymandr commented:
> * QC: Uncaught TypeError: som.Spinner is not a constructor
Fixed

> * Fixtures: Uncaught TypeError: som.setsharedvalue is not a function
Fixed

> * Fixtures: Uncaught TypeError: _jquery2.default(...)[0] is undefined attempting to detect markers on fixture
Not sure how to reproduce.

> * Feature extract: Erasing `/root` text in analysis folder causes console error
Cannot reproduce.

> *Analysis: No message on submit analysis (not chained)
Cannot reproduce (or I don't understand the problem – I get a popup and am redirected to status)

> * Compile/General: Console has a lot of errors
I have fixed a lot of those.

> * Compile: Noting happens on submit
Cannot reproduce (or I don't understand the problem – I get a popup)

> * Experiment/Scan: Doesn't show CCC calibration, analysis page does so should not be an issue
Cannot reproduce (or I don't understand the problem – it is shown for me, it seems)

Confirmed/To do:
- [ ] General: CSS
- [ ] Experiment/Scan: focus to Description field cause console error
- [ ] Experiment/Scan: Age change causes console errors
- [ ] Experiment/Scan: Time from pinning to experiment causes console errors

local-minimum commented on scanomatic/ui_server_data/compile.html:
```diff
@@ -80,7 +80,7 @@ <h3>Chaining</h3>
+    som.setSharedValue('imageListDiv', $('#manual-selection-div'));
     manual_btn = $('#manual-selection');
     som.InputEnabled(manual_btn, false);
     som.toggleManualSelectionBtn(manual_btn);
```
```suggestion
    manualBtn = $('#manual-selection');
    som.InputEnabled(manualBtn, false);
    som.toggleManualSelectionBtn(manualBtn);
```

local-minimum reviewed (commented):
I'll have a look running the code when you are ready.

local-minimum commented on scanomatic/ui_server_data/analysis.html:
```diff
@@ -2,17 +2,12 @@
-  <script src="/js/fixtures.js?ver=_-_VERSIONTAG_-_"></script>
-  <script src="/js/analysis.js?ver=_-_VERSIONTAG_-_"></script>
-  <script src="/js/calibration.js?ver=_-_VERSIONTAG_-_"></script>
+  <script src="/js/somlib/som.js?ver=_-_VERSIONTAG_-_"></script>
```
here you have the version tag thing on `som.js`, but in the two previous files it's not. Don't know if it's needed but I think it at least should be systematic.

skymandr replied:
I have retained it where it was originally, but some files lacked it. I didn't want to change this, in case there was a thought behind this, but I can add or remove it everywhere if you want. Note that one file (`qc_norm.html`) sets version explicitly to 1.0.0. Anyway, let me know what you want me to do about this.

local-minimum replied:
if the ccc file import which is made in the same way had it, then add it to all, if the ccc file doesn't then remove it from all som.js

skymandr replied:
`ccc.js` does not have version, so I will remove it from all.

skymandr replied:
I noted that the css-links are also inconsistently handled, some have the version tags, others do not. I suggest I remove them while I'm at it, since I have to go through all the files anyway.

skymandr replied:
fixed (css are now included in #128)

local-minimum commented on scanomatic/ui_server_data/js/som/analysis.js:
```diff
@@ -1,15 +1,41 @@
-  set_fixture_plate_listing();
+function appendRegriddingUI(parent, plateIndex) {
+  parent.append((
+    `<div class='plate-regridding' id='plate-regridding-${plateIndex}' onmouseleave='hideGridImage();'>`
+      + '<fieldset>'
+      + `<img class='grid_icon' src='/images/grid_icon.png' onmouseenter='loadGridImage(${plateIndex - 1});'>`
+      + `<legend>Plate ${plateIndex}</legend>`
+
+      + `<input type='radio' name='plate-regridding-radio-${plateIndex}' value='Keep' checked='checked'>`
+      + `<label id='plate-regridding-keep${plateIndex}'>Keep previous</label><br>`
+
+      + `<input type='radio' name='plate-regridding-radio-${plateIndex}' value='Offset'>`
+      + `<label id='plate-regridding-offset${plateIndex}'>Offset</label>`
+      + `<input type='number' class='plate-offset' id='plate-regridding-offset-d1-${plateIndex}' value='0' name='Offset-d1'>`
+      + `<input type='number' class='plate-offset' id='plate-regridding-offset-d2-${plateIndex}' value='0' name='Offset-d2'><br>`
+
+      + `<input type='radio' name='plate-regridding-radio-${plateIndex}' value='New'>`
+      + `<label id='plate-regridding-new${plateIndex}'>New grid from scratch</label><br>`
+      + '</fieldset>'
+    + '</div>'
```
does this really work, these callbacks in text form like `hideGridImage()` in the first line are part of the `som` namespace right so it should be `som.hideGridImage();` and so on.

Also check `<label id='plate-regridding-offset${plateIndex}'>Offset</label>` and `<label id='plate-regridding-new${plateIndex}` if it really should be without a dash before plateIndex as it is in all other cases.

skymandr replied:
Yes, `som.` is needed. Well spotted.

As for the id's I'll try to have a look.

skymandr replied:
Fixed. The ids aren't actually referenced anywhere, so I've added `-`.

local-minimum replied:
should be quite searchable with regridding-new and regridding-offset

local-minimum commented on scanomatic/ui_server_data/js/som/experiment.js:
```diff
@@ -0,0 +1,376 @@
+        "<input type='hidden' value='"}${index}'>` +
+        `<label for='plate-description-${index}'>Plate ${index}</label>` +
+        `<input class='long' id='plate-description-${index}' value='${description
+        }' placeholder='Enter medium and, if relevant, what experiment this plate is.' onchange='cacheDescription(this);'></div>`;
```
this onchange should probably be `som.cache...`

skymandr replied:
Fixed.

local-minimum commented on scanomatic/ui_server_data/js/som/experiment.js:
```diff
@@ -0,0 +1,376 @@
+  return (
+    `<div class='pinning'><input type='hidden' value='${plate.index}'>`
+      + `<label for='pinning-plate-${plate.index}'>Plate ${plate.index}</label>`
+      + `<select id='pinning-plate-${plate.index}' class='pinning-selector' onchange='setActivePlate(this); validateExperiment();'>`
```
this onchange too

local-minimum replied:
basically check all onX occurances

skymandr replied:
I checked all `onclick` yesterday, but forgot about `onchange`. Thanks.

skymandr replied:
Fixed (checked all ` on[a-z]+`)

local-minimum replied:
onmouse onkey ondocument perhaps we have others too

local-minimum commented on scanomatic/ui_server_data/js/som/qc_normAPIHelper.js:
```diff
@@ -0,0 +1,399 @@
+import { getLastSegmentOfPath } from './qc_normHelper';
+import { setSharedValue } from './helpers';
+
+// var baseUrl = "http://localhost:5000";
```
remove?

skymandr replied:
Fixed.

local-minimum commented on scanomatic/ui_server_data/js/som/qc_normDrawPlate.js:
```diff
@@ -169,7 +178,62 @@ function addSelectionHanddling(svgRoot) {
+  const legendWidth = 25;
+  const legendMargin = 5;
+
+  // SetDebugText(data, cols, rows);
```
```suggestion
```

skymandr replied:
Should I also remove `SetDebugText`?

local-minimum replied:
if its not used yeah remove it

skymandr replied:
Done.

local-minimum commented on scanomatic/ui_server_data/js/som/qc_normDrawPlate.js:
```diff
@@ -169,7 +178,62 @@ function addSelectionHanddling(svgRoot) {
+  heatMap.displayLegend(true);
+  heatMap.dispatch2(dispatch);
+  heatMap(plateGroup);
+  // heatMap.on(dispatcherSelectedExperiment, function (datah) {
+  //    console.log("dispatched:" + datah);
+  // });
```
```suggestion
```

skymandr replied:
Fixed.

local-minimum commented on scanomatic/ui_server_data/qc_norm.html:
```diff
@@ -3,40 +3,19 @@
-    <script src="js/qc_normDrawCurves.js?ver=1.0.0"></script>
-    <script src="js/qc_normAPIHelper.js?ver=1.0.0"></script>
-
+    <script src="js/somlib/som.js?ver=1.0.0"></script>
```
why this ver here?

skymandr replied:
You tell me!

local-minimum replied:
I have no clue

skymandr replied:
I'll just remove it, as you suggested above then.

skymandr replied:
Fixed, but see comment up top.

local-minimum reviewed (changes requested):
Gone through the code again and found some more details.

local-minimum commented:
I've found what I've managed to spot, any issues remaining we'll have to deal with as separate bugs.
